From edf9852627422764eb0b90cd1ba21e43c6f424c1 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Wed, 26 Jun 2024 14:25:26 +0200
Subject: [PATCH 0980/1695] plat-stm32mp1: rework PMIC usages

Rework the way the PMICs are used by the platform.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0714fce7f7a3847e5eec4a12abeedd5b902e01d0 plat-stm32mp1: rework PMIC usages]
Change-Id: Ic833e67cc10ccc69c37fcc45d9e7e7b12a952321
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/389285
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 .../arm/plat-stm32mp1/drivers/stm32mp1_pmic.c |  2 +-
 core/arch/arm/plat-stm32mp1/main.c            |  5 ----
 core/arch/arm/plat-stm32mp1/pm/low_power.c    |  7 +++---
 core/arch/arm/plat-stm32mp1/stm32_util.h      | 25 +++++++++++++++++--
 core/arch/arm/plat-stm32mp2/main.c            |  2 +-
 core/arch/arm/plat-stm32mp2/stm32_util.h      |  4 +--
 core/drivers/stm32mp_pmic2.c                  |  6 ++---
 7 files changed, 34 insertions(+), 17 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
index faf91b10b..707021e25 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c
@@ -84,7 +84,7 @@ static bool pmic_is_secure(void)
 	return pmic_status == DT_STATUS_OK_SEC;
 }
 
-bool stm32mp_with_pmic(void)
+bool stm32_stpmic1_is_present(void)
 {
 	return pmic_status != -1;
 }
diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index 914cf155c..7f8de4da2 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -433,11 +433,6 @@ void plat_bsec_get_static_cfg(struct stm32_bsec_static_cfg *cfg)
 	cfg->max_id = STM32MP1_OTP_MAX_ID;
 }
 
-bool __weak stm32mp_with_pmic(void)
-{
-	return false;
-}
-
 uint32_t may_spin_lock(unsigned int *lock)
 {
 	if (!lock || !cpu_mmu_enabled())
diff --git a/core/arch/arm/plat-stm32mp1/pm/low_power.c b/core/arch/arm/plat-stm32mp1/pm/low_power.c
index 78d976f50..9f8026e37 100644
--- a/core/arch/arm/plat-stm32mp1/pm/low_power.c
+++ b/core/arch/arm/plat-stm32mp1/pm/low_power.c
@@ -14,7 +14,6 @@
 #include <drivers/stm32_etzpc.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <drivers/stm32mp1_ddrc.h>
-#include <drivers/stm32mp1_pmic.h>
 #include <drivers/stm32mp1_pwr.h>
 #ifdef CFG_STM32MP13
 #include <drivers/stm32mp13_rcc.h>
@@ -229,7 +228,8 @@ void stm32_enter_cstop(uint32_t mode)
 				    mode == STM32_PM_CSTOP_ALLOW_LPLV_STOP2))
 		pwr_cr1 |= PWR_CR1_LPCFG;
 
-	if (stm32mp_with_pmic()) {
+	/* The PMIC2 applies it's low power config in a suspend call back */
+	if (stm32_stpmic1_is_present()) {
 		stm32mp_get_pmic();
 		stm32mp_pmic_apply_lp_config(config_pwr[mode].regul_suspend_node_name);
 		stm32mp_put_pmic();
@@ -367,7 +367,8 @@ void __noreturn stm32_enter_cstop_shutdown(uint32_t mode)
 
 	switch (mode) {
 	case STM32_PM_SHUTDOWN:
-		if (stm32mp_with_pmic()) {
+		/* The PMIC2 driver does not implement switch off function */
+		if (stm32_stpmic1_is_present()) {
 			wait_console_flushed();
 			stm32mp_get_pmic();
 			stpmic1_switch_off();
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index faa1fd63c..2efc1616c 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -9,6 +9,7 @@
 #include <assert.h>
 #include <drivers/clk.h>
 #include <drivers/pinctrl.h>
+#include <drivers/stm32mp1_pmic.h>
 #include <drivers/stm32mp1_rcc_util.h>
 #include <kernel/panic.h>
 #include <stdbool.h>
@@ -44,8 +45,28 @@ TEE_Result stm32mp_syscfg_erase_sram3(void);
 /* Platform util for the GIC */
 vaddr_t get_gicd_base(void);
 
-/* Platform util for PMIC support */
-bool stm32mp_with_pmic(void);
+#ifdef CFG_STPMIC1
+bool stm32_stpmic1_is_present(void);
+#else
+static inline bool stm32_stpmic1_is_present(void)
+{
+	return false;
+}
+#endif
+
+#ifdef CFG_STPMIC2
+bool stm32_stpmic2_is_present(void);
+#else
+static inline bool stm32_stpmic2_is_present(void)
+{
+	return false;
+}
+#endif
+
+static inline bool stm32mp_with_pmic(void)
+{
+	return stm32_stpmic1_is_present() || stm32_stpmic2_is_present();
+}
 
 /* Power management service */
 #ifdef CFG_PSCI_ARM32
diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index 98e0dce0f..b77de50a8 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -499,7 +499,7 @@ void stm32_debug_suspend(unsigned long a0)
 
 	/* Pulse on NRSTC1MS pin for standby request when PMIC is used */
 	if (dbgmcu_cr & DBGMCU_CR_DBG_STANDBY && a0 >= PM_D2_LPLV_LEVEL &&
-	    stm32_pmic2_is_present())
+	    stm32_stpmic2_is_present())
 		io_setbits32(stm32_rcc_base() + RCC_C1MSRDCR,
 			     RCC_C1MSRDCR_C1MSRST);
 
diff --git a/core/arch/arm/plat-stm32mp2/stm32_util.h b/core/arch/arm/plat-stm32mp2/stm32_util.h
index c86091730..5c73d95a7 100644
--- a/core/arch/arm/plat-stm32mp2/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp2/stm32_util.h
@@ -33,9 +33,9 @@ bool stm32mp_supports_cpu_opp(uint32_t opp_id);
 #endif /*CFG_STM32_CPU_OPP*/
 
 #ifdef CFG_STPMIC2
-bool stm32_pmic2_is_present(void);
+bool stm32_stpmic2_is_present(void);
 #else
-static inline bool stm32_pmic2_is_present(void)
+static inline bool stm32_stpmic2_is_present(void)
 {
 	return false;
 }
diff --git a/core/drivers/stm32mp_pmic2.c b/core/drivers/stm32mp_pmic2.c
index 73a5fc82c..c12deb8ff 100644
--- a/core/drivers/stm32mp_pmic2.c
+++ b/core/drivers/stm32mp_pmic2.c
@@ -962,9 +962,9 @@ DEFINE_DT_DRIVER(stm32_pmic2_dt_driver) = {
 	.match_table = stm32_pmic2_match_table,
 	.probe = stm32_pmic2_probe,
 };
-#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23) || defined(CFG_STM32MP21)
-bool stm32_pmic2_is_present(void)
+
+bool stm32_stpmic2_is_present(void)
 {
 	return stm32_pmic2;
 }
-#endif
+
-- 
2.47.2

